contrib/python/podman/libs/__init__.py

```diff
@@ -58,29 +58,35 @@ def datetime_parse(string):
     # match: 2018-05-08T18:24:52.753227-07:00
     # match: 2018-05-08 14:12:53.797795191 -0700 MST
     # match: 2018-05-09T10:45:57.576002  (python isoformat())
+    # match: 2018-05-17T16:32:22.326956313Z
 
     Some people, when confronted with a problem, think “I know,
     I'll use regular expressions.”  Now they have two problems.
       -- Jamie Zawinski
     """
     ts = re.compile(r'^(\d+)-(\d+)-(\d+)'
                     r'[ T]?(\d+):(\d+):(\d+).(\d+)'
-                    r' *([-+][\d:]{4,5})? *')
+                    r'([ zZ]*)([-+][\d:]{4,5})? *')
```

Comment:
This still has the the `Z` entry in a different group from the offsets.  I'd [rather have them in the same group][1].  For example, this regexp will match:

    2018-05-17T16:32:22.326956313 Z ZZZ Z +01:23

and I don't think we want to allow that.  Although without a trailing `$`, [my regexp fragment][1] will match that too just without the offset in a group.  So I'll add a `$` to my recommendation and make it:

```python
r' *([zZ]|[-+][\d:]{4,5})? *$'
```

[1]: https://github.com/projectatomic/libpod/pull/800#discussion_r189165543

Reply:
@wking I agreed I'm letting you get crazy with Z's.  But simply adding a trailing $ still won't parse 2018-05-08 14:12:53.797795191 -0700 MST

Reply:
> But simply adding a trailing $ still won't parse 2018-05-08 14:12:53.797795191 -0700 MST

I'd rather address that directly then.  Using [`(?P<name>` and `groupdict`][1], you could have:

```python
ts = re.compile(r'^(?P<year>\d+)-(?P<month>\d+)-(?P<day>\d+)'
                     r'[ T]?(?P<hour>\d+):(?P<minute>\d+):(?P<second>\d+(.\d*)?)'
                     r' *(?P<zone>[zZ]|(?P<offset>[-+][\d:]{4,5})(?: *[a-zA-Z]+)?)? *$')
```

That would give:

```pycon
>>> for string in [
...         '2018-05-08T14:12:53.797795191-07:00',
...         '2018-05-08T18:24:52.753227-07:00',
...         '2018-05-08 14:12:53.797795191 -0700 MST',
...         '2018-05-09T10:45:57.576002',
...         '2018-05-17T16:32:22.326956313Z',
...         ]:
...     print(ts.match(string).groupdict())
... 
{'year': '2018', 'month': '05', 'day': '08', 'hour': '14', 'minute': '12', 'second': '53.797795191', 'zone': '-07:00', 'offset': '-07:00'}
{'year': '2018', 'month': '05', 'day': '08', 'hour': '18', 'minute': '24', 'second': '52.753227', 'zone': '-07:00', 'offset': '-07:00'}
{'year': '2018', 'month': '05', 'day': '08', 'hour': '14', 'minute': '12', 'second': '53.797795191', 'zone': '-0700 MST', 'offset': '-0700'}
{'year': '2018', 'month': '05', 'day': '09', 'hour': '10', 'minute': '45', 'second': '57.576002', 'zone': None, 'offset': None}
{'year': '2018', 'month': '05', 'day': '17', 'hour': '16', 'minute': '32', 'second': '22.326956313', 'zone': 'Z', 'offset': None}
```

[1]: https://docs.python.org/3.5/library/re.html#re.match.groupdict